opendatahub-io/data-science-pipelines
# UPSTREAM: <carry>: add last_run_creation

**Description of your changes:**

Resolves: https://issues.redhat.com/browse/RHOAIENG-7692

### Testing instructions: 

<details> 

<summary>instruction.txt</summary>

#### Prepare environment: 
1. First deploy regular DSPA without changes. 
2. Create experiments `expA` and `expB` 
3. Create a `run1A` and `run1B` in `expA` and `expB`  respectively
4. Get access to the underlying mariadb instance, you will need to be able to `select` on run table and experiment table
Go to mariadb pod, open the terminal 
```
mysql -u root 
MariaDB [(none)]> use mlpipeline
Database changed
MariaDB [mlpipeline]> select * from experiments;
```

Confirm that you don't see a column `LastRunCreatedAtInSec` .  

> Note: you can also use [cloudbeaver for this. ](https://github.com/HumairAK/cloudbeaver-k8s)

#### Now deploy the changes: 
Update the current DSPA with the updated images, you should only need to update the API server image


#### Test Runs: 
1. Now create a run named `run2A`, `run3A`, `run4A` in `expA` experiment, **make sure the runs are not cached.**
You can use this sample pipeline: [here](https://gist.github.com/HumairAK/3ccfdade7f24e8896a6629e4075d6143)

2. Now either using postman or curl, the experiment and sort by `last_run_created_at`: 

```
curl --location --request GET '${API_SERVER_ROUTE}/apis/v2beta1/experiments?namespace=${YOUR_DSPA_NS}&sort_by=last_run_created_at desc' \
--header 'Authorization: Bearer ${YOUR_OC_AUTH_TOKEN}'
```
Set `API_SERVER_ROUTE`, `YOUR_OC_AUTH_TOKEN`, `YOUR_DSPA_NS` accordingly. 

You should see `expA` and `expB` listed in a json sorted by `last_run_created_at`. Confirm the presence of this field `last_run_created_at` in ***every*** experiment listed.

Note that after deploying we did not create a new run in `expB`, thus we expect `expB`'s `last_run_created_at` to be the first epoch time: `"last_run_created_at": "1970-01-01T00:00:00Z"` confirm this is the case. This should confirm the update scenario for existing experiments. Because of this since we are sorting by `desc` we expect `expB` to show up last. Now try to use `asc` instead of `dsc` confirm the order is reversed. You can repeat this test with multiple experiments instead of just 2. 

We also need to test the behavior with scheduled runs because these are persisted by persistence Agent so they work differently.

#### Test Jobs: 

1. Create a recurring run `run2B` under `expB`. Set the cron for every 1 min.
2. Let 3 minutes pass, or until 3 runs are created. Disable the recurring run via the UI for `run2B` otherwise you'll get spammed with runs. 
3. Now use curl or postman again, the command is the same as before, and sort by `last_run_created_at`. Confirm expected behavior.
4. `expB` should not show first if sorting by `desc` and the `last_run_created_at` should match this run's `creation time` (see this in the runs page, or curl the runs api)
5. Confirm `asc` sorting has expected behavior


#### Create a new experiment

1. Create a new experiment `exp3` after you have deployed the new changes. 
2. Once again do a sort, `exp3` should show up last when sorting by `desc` because it has no runs. Confirm that it has the first epoch time stamp.
3. Create a run in `exp3`, and curl the api again and sort by `last_run_created_at` confirm expected behavior, sort again
#### Verify DB data

Now go to mariadb, as described earlier, display the experiments table, confirm `LastRunCreatedAtInSec` is populated as expected this maps to `last_run_created_at`. 

For converting epoch, [you can use this tool for relative time](https://www.epochconverter.com/)

</details>
